lib/lookup/src/lookup_v2/mod.rs

```diff
@@ -220,3 +261,56 @@ impl<'a> From<BorrowedSegment<'a>> for OwnedSegment {
         }
     }
 }
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn owned_path_display() {
+        let test_cases = [
+            ("", ".<invalid>"),
+            (".", "."),
+            ("]", ".<invalid>"),
+            ("]foo", ".<invalid>"),
+            ("..", ".<invalid>"),
+            ("...", ".<invalid>"),
+            ("f", ".f"),
+            (".f", ".f"),
+            (".[", ".<invalid>"),
+            ("f.", ".f"), // TODO: Review, shouldn't this be invalid?
+            ("foo", ".foo"),
+            (
+                r#"ec2.metadata."availability-zone""#,
+                r#".ec2.metadata."availability-zone""#,
+            ),
+            (".foo", ".foo"),
+            (".@timestamp", ".@timestamp"),
+            ("foo[", ".foo.<invalid>"),
+            ("foo$", ".<invalid>"),
+            (r#""$peci@l chars""#, r#"."$peci@l chars""#),
+            (".foo.foo bar", ".foo.<invalid>"),
+            (r#".foo."foo bar".bar"#, r#".foo."foo bar".bar"#),
+            ("[1]", "[1]"),
+            ("[42]", "[42]"),
+            (".[42]", ".<invalid>"),
+            ("[42].foo", "[42].foo"),
+            ("[42]foo", "[42].foo"), // TODO: Review, shouldn't this be invalid?
```

Comment:
Agreed, good to file as separate issue.

Reply:
Not sure I agree with this one. `[42]foo` is not ambiguous what it represents, and can easily be parsed. It is also currently supported in VRL.

Reply:
That's funny. Then we probably should not break backwards compatibility here?

Reply:
Is that supported since VRL uses the same lookup code now, or has that also been supported pre `0.21`?

Reply:
It has been supported in VRL since at least 0.19.0, and probably much longer

Reply:
Thanks for confirming 👍

Reply:
It has, but it's a bug. We will definitely fix this at some point, but we don't need to do that here, I agree.

Reply:
Also mentioned this case in #12326